src/lib/modules/deployment/contract_deployer.js

```diff
@@ -209,32 +209,46 @@ class ContractDeployer {
       function doLinking(next) {
+
+        if (!contract.linkReferences || !Object.keys(contract.linkReferences).length) {
+          return next();
+        }
         let contractCode = contract.code;
-        self.events.request('contracts:list', (_err, contracts) => {
-          for (let contractObj of contracts) {
-            let filename = contractObj.filename;
-            let deployedAddress = contractObj.deployedAddress;
-            if (deployedAddress) {
-              deployedAddress = deployedAddress.substr(2);
-            }
-            let linkReference = '__' + filename + ":" + contractObj.className;
-            if (contractCode.indexOf(linkReference.substr(0, 38)) < 0) { // substr to simulate the cut that solc does
-              continue;
-            }
-            if (linkReference.length > 40) {
-              return next(new Error(__("{{linkReference}} is too long, try reducing the path of the contract ({{filename}}) and/or its name {{contractName}}", {linkReference: linkReference, filename: filename, contractName: contractObj.className})));
-            }
-            let toReplace = linkReference + "_".repeat(40 - linkReference.length);
-            if (deployedAddress === undefined) {
-              let libraryName = contractObj.className;
-              return next(new Error(__("{{contractName}} needs {{libraryName}} but an address was not found, did you deploy it or configured an address?", {contractName: contract.className, libraryName: libraryName})));
-            }
-            contractCode = contractCode.replace(new RegExp(toReplace, "g"), deployedAddress);
-          }
-          // saving code changes back to the contract object
+        let offset = 0;
+
+        async.eachLimit(contract.linkReferences, 1, (fileReference, eachCb1) => {
+          async.eachOfLimit(fileReference, 1, (references, libName, eachCb2) => {
+            self.events.request("contracts:contract", libName, (libContract) => {
+              async.eachLimit(references, 1, (reference, eachCb3) => {
+                if (!libContract) {
+                  return eachCb3(new Error(__('{{contractName}} has a link to the library {{libraryName}}, but it was not found. Is it in your contract folder?'), {
+                    contractName: contract.className,
+                    libraryName: libName
+                  }));
+                }
+
+                let libAddress = libContract.deployedAddress;
+                if (!libAddress) {
+                  return eachCb3(new Error(__("{{contractName}} needs {{libraryName}} but an address was not found, did you deploy it or configured an address?", {
+                    contractName: contract.className,
+                    libraryName: libName
+                  })));
+                }
+
+                libAddress = libAddress.substr(2).toLowerCase();
+
+                // Multiplying by two because the original pos and length are in bytes, but we have an hex string
```

Comment:
Even with the comment I am not sure to understand exactly why the `* 2`
Do you have math example that could help me?

Reply:
Example, if the reference says : {start: 30, length: 20}, the real position of the `__lib.sol:whatever___` will be 60, and the length will be 40 (the same as an eth address)
That is because they setup the position and length from the byte count, but the string we get is in hexadecimal. 
A byte is 8 bits, whereas an hexa is 16, hence the times 2